bolt-contracts/src/contracts/BoltChallenger.sol

```diff
@@ -0,0 +1,357 @@
+    /// enough time has passed without proper resolution.
+    ///
+    /// q: should we also have a commit-reveal scheme for resolutions to avoid frontrunning to steal bonds?
+    function _resolve(bytes32 challengeID, bytes32 trustedBlockHash, Proof calldata proof) internal {
+        // The challenge is assumed to exist at this point, so we can safely access it.
+        Challenge storage challenge = challenges[challengeID];
+
+        if (challenge.status != ChallengeStatus.Open) {
+            revert ChallengeAlreadyResolved();
+        }
+
+        if (challenge.openedAt + MAX_CHALLENGE_DURATION < Time.timestamp()) {
+            // If the challenge has expired without being resolved, it is considered lost.
+            challenge.status = ChallengeStatus.Lost;
+            _transferFullBond(challenge.challenger);
+            emit ChallengeLost(challengeID);
+            return;
+        }
+
+        // Verify the validity of the header against the trusted block hash.
+        if (keccak256(proof.blockHeaderRLP) != trustedBlockHash) {
+            revert InvalidBlockHash();
+        }
+
+        // Decode the RLP-encoded block header fields
+        BlockHeaderData memory blockHeader = _decodeBlockHeaderRLP(proof.blockHeaderRLP);
+
+        // Decode the committed raw signed transaction. Its sender will be the account to prove existence of.
+        TransactionDecoder.Transaction memory decodedTx = challenge.commitment.signedTx.decodeEnveloped();
+
+        // Decode the account fields by checking the account proof against the state root of the block header.
+        // The key in the account trie is the account pubkey (address) which we can recover from the signed tx.
+        (bool accountExists, bytes memory accountRLP) = SecureMerkleTrie.get(
+            abi.encodePacked(decodedTx.recoverSender()), proof.accountMerkleProof, blockHeader.stateRoot
+        );
+
+        if (!accountExists) {
+            revert AccountDoesNotExist();
+        }
+
+        AccountData memory account = _decodeAccountRLP(accountRLP);
+
+        if (account.nonce > decodedTx.nonce) {
+            // The transaction recovered sender has sent a transaction with a higher nonce than the committed
+            // transaction, before the proposer could include it. Consider the challenge defended, as the
+            // proposer is not at fault. The bond will be shared between the resolver and commitment signer.
+            challenge.status = ChallengeStatus.Defended;
+            _transferHalfBond(msg.sender);
+            _transferHalfBond(challenge.target);
+            emit ChallengeDefended(challengeID);
+            return;
+        } else if (account.nonce < decodedTx.nonce) {
+            // Q: is this a valid case? technically the proposer would be at fault for accepting a commitment of an
+            // already included transaction. TBD.
+        }
+
+        if (account.balance < blockHeader.baseFee * decodedTx.gasLimit) {
```

Comment:
Kinda related to the nonce check: if the account state considered _after all the transactions have been applied_? If so, this check might return true because the user might have consumed all its balance even if it has actually paid for the preconf. 
If the account state is _before transactions have been applied_, could you please document it very clearly? This has a great impact on the dispute logic.

Reply:
You're right, the current challenge doesn't handle stacked preconfs in the same slot.
This will need more attention, moving convo offline

Reply:
Addressed in #239